src/transformers/models/falcon/modeling_falcon.py

```diff
@@ -209,6 +209,162 @@ def _set_cos_sin_cache(self, seq_len, device, dtype):
+        device=None,
+    ):
+        super().__init__(dim, max_position_embeddings, base, device)
+
+        self.scaling_factor = scaling_factor
+        self.original_max_position_embeddings = original_max_position_embeddings
+        self.extrapolation_factor = extrapolation_factor
+        self.attention_factor = attention_factor
+        self.beta_fast = beta_fast
+        self.beta_slow = beta_slow
+
+        self.yarn(device)
+
+        # Build here to make `torch.jit.trace` work.
+        self.max_seq_len_cached = max_position_embeddings
+        emb = self.get_pos_embeddings(device)
+
+        self._cos_cached = (emb.cos() * self.mscale)[:, :].to(torch.get_default_dtype())
+        self._sin_cached = (emb.sin() * self.mscale)[:, :].to(torch.get_default_dtype())
+
+    # Get positional embeddings based on the current max sequence length
+    def get_pos_embeddings(self, device):
+        t = torch.arange(self.max_seq_len_cached, device=device, dtype=self.inv_freq.dtype)
+        freqs = torch.einsum("i,j->ij", t, self.inv_freq)
+        emb = torch.cat((freqs, freqs), dim=-1)
+        return emb
+
+    # Inverse dimension formula to find the dimension based on the number of rotations
+    def find_correction_dim(self, num_rotations, dim, base=10000, max_position_embeddings=2048):
+        return (dim * math.log(max_position_embeddings / (num_rotations * 2 * math.pi))) / (2 * math.log(base))
+
+    # Find dimension range bounds based on rotations
+    def find_correction_range(self, low_rot, high_rot, dim, base=10000, max_position_embeddings=2048):
+        low = math.floor(self.find_correction_dim(low_rot, dim, base, max_position_embeddings))
+        high = math.ceil(self.find_correction_dim(high_rot, dim, base, max_position_embeddings))
+        return max(low, 0), min(high, dim - 1)
+
+    def linear_ramp_mask(self, min, max, dim):
+        if min == max:
+            max += 0.001  # Prevent singularity
+
+        linear_func = (torch.arange(dim, dtype=torch.float32) - min) / (max - min)
+        ramp_func = torch.clamp(linear_func, 0, 1)
+        return ramp_func
+
+    def get_mscale(self, scaling_factor=1):
+        if scaling_factor <= 1:
+            return 1.0
+        return 0.1 * math.log(scaling_factor) + 1.0
+
+    def forward(self, x, seq_len=None):
+        # Difference to the original RoPE: applies a scaling factor computed with
+        # the YaRN method (NTK-by-Parts + Attn Scaling)
+        # x: [batch_size, seq_len, head_dim]
+        if seq_len > self.max_seq_len_cached:
+            self.max_seq_len_cached = seq_len
+            emb = self.get_pos_embeddings(x.device)
+
+            self._cos_cached = (emb.cos() * self.mscale)[:, :].to(x.dtype)
+            self._sin_cached = (emb.sin() * self.mscale)[:, :].to(x.dtype)
+        return (
+            self._cos_cached[:seq_len, ...].to(dtype=x.dtype),
+            self._sin_cached[:seq_len, ...].to(dtype=x.dtype),
+        )
+
+    def yarn(self, device):
```

Comment:
```suggestion
    def compute_yarn_scaling(self, device):
```
(Or a similar name. Let's use descriptive function names :) )